[IMP] survey: add language selector

pyduf opened the pull request:
We add a language selector in the first page of the survey allowing users to select their language. This selector is only present if more than one language is enabled on the survey and it is removed once the survey begins.

The survey form allows the creator of a survey to configure the available languages among the installed ones. By default, the context language is enabled. If no language is selected for the survey, it means that the survey supports all installed languages. As the demo surveys support multiple language, we left that field empty for them.

When resuming a survey, the language is restored. To support that feature, we now save the user language for each survey_input which allow us also to add the following features:
- add a filter allowing survey reviewer to filter survey/answer based on the survey language.
- print survey in the user language

Technical notes:
- we can't use the language selector of portal as survey doesn't depend on portal.
- to avoid adding new tours, we test the language selector by modifying the public tour. We add a step to switch to French and verify that this language is used during the whole survey.
- we have decided to add the field lang_ids on survey after having tried to auto-detect in which languages a survey is translated. We have given up on that solution because to be reliable it required to use a sql query and it was a bit too magical for the user (the language was enabled as soon as one translation was available as not everything must always be translated. Ex: numbers).

[IMP] website, survey: avoid limiting survey to website language

Without the module website, the survey can be displayed in any language installed but as soon as website is installed, the survey is limited to the main website languages. This commit allows to display a survey in any installed language even if not all installed languages are enabled in the main website.

Task-3884012

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/165760).

lm4649 commented:
lgtm @pyduf!

flch-odoo commented on addons/survey/static/tests/tours/survey.js:
```diff
@@ -76,4 +75,53 @@ registry.category("web_tour.tours").add('test_survey', {
+        },
+    },
+    {
+        content: "Check French translation",
```
Shouldn't we instead/also translate one question and switch the step for that question, so that we check that we correctly "remember" the selected language throughout the "session"?

flch-odoo commented on addons/survey/controllers/main.py:
```diff
@@ -389,6 +393,9 @@ def _prepare_question_html(self, survey_sudo, answer_sudo, **post):
 
     @http.route('/survey/<string:survey_token>/<string:answer_token>', type='http', auth='public', website=True)
     def survey_display_page(self, survey_token, answer_token, **post):
+        if (lang_code := post.get('lang_code')) and request.httprequest.method == 'POST':
+            return request.redirect(self.env['ir.http']._url_for(f'/survey/{survey_token}/{answer_token}', lang_code))
```
I'm not sure I see the benefit of this server-side manipulation...
Why not simply redirect in JS with a `onChangeLanguage` handler on the `select`?

(and let's rename this `**post` into `**kwargs` either way if the route supports GET too, no?)

pyduf reviewed (commented):
Hello @flch-odoo , 

Thanks for the review. I have updated the code and answered your comments.

flch-odoo commented on addons/survey/static/src/js/survey_form.js:
```diff
@@ -46,7 +46,7 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend(SurveyPreloa
         this.fadeInOutDelay = 400;
         return this._super.apply(this, arguments).then(function () {
-            self.options = self.$('form').data();
+            self.options = self.$("form.o_survey-fill-form").data();
```
I understand. Thank you for the research on this. If there is no 1-to-1 replacement with vanilla/browser JS, then it should be handled differently, not only for our case here.

We should have a utility function to do that (ideally one for Odoo) and not have it inline here.

If you think it's "robust" enough for our use case, please make it a function in a utils file, where you can add the fadeOut as well (it shouldn't be a SurveyWiget method).

As you said, it is not our priority either, so if's too much work for now, it's OK to keep jQuery here for now.

flch-odoo reviewed (commented):
Thanks for the update, @pyduf !

Food for thoughts on the next bright new day (or next after that) :)

flch-odoo commented on addons/survey/models/survey_survey.py:
```diff
@@ -1258,6 +1258,34 @@ def _create_certification_badge_trigger(self):
+        """
+        self.ensure_one()
+        langs = self.env['res.lang']._get_frontend()
+        lang_base = 'en_US'
```
I think you can use the first language from `langs` instead, no? 
If all translations are equal to this, we'll show no option, if some differ, this first lang will be a valid option.
The hardcoded "en_US" part is an implementation of the ORM for the db record, we don't need to explicitly use it.

pyduf replied:
Yes, maybe, it would be easier to get rid of this auto-detection or accept it may make some mistakes as I haven't found a way by not using SQL nor en_US as base language to "We trust those frontend languages (but we avoid showing selector if nothing was translated)". As if 2 languages have the same translation, we cannot know which one must be included or even if none of them must be.
What do you think ? Do you have a better idea ?

flch-odoo replied:
Yes :) That, or explicitly setting available translations on the survey, following PO's preference.

pyduf reviewed (commented):
Hello @flch-odoo , 

Thanks for the review. I have updated the code and answered your comments.

flch-odoo commented on addons/website/models/res_lang.py:
```diff
@@ -15,15 +15,19 @@ def write(self, vals):
-            lang_ids = self.env['website'].get_current_website().with_context(lang=False).language_ids.sorted('name').ids
-            langs = [dict(self.env['res.lang']._get_data(id=id_)) for id_ in lang_ids]
+            if self.env.context.get('web_force_installed_langs'):
+                langs = sorted([dict(data) for data in self._get_active_by('code').values()],
+                               key=lambda lang: lang['name'])
+            else:
+                lang_ids = self.env['website'].get_current_website().with_context(lang=False).language_ids.sorted('name').ids
```
Can't we just define `lang_ids` and leave the creation of `langs` as it was for both cases?

flch-odoo reviewed (commented):
Good work, @pyduf ! One or two validation comments left on my end :)

pyduf reviewed (commented):
Hello @flch-odoo , 

Thanks for the review, I have answered your comments.
I can add a way to remember the language if it is interrupted but there is something I am missing about your comment (see below).

pyduf commented:
Hello @flch-odoo , 

Following our discussion, I have added the language on survey_input and the related features (restore the language when resuming a survey, filter on view based on this field, print the survey in the language in which the survey was filled). Note I haven't attempted to print the certification in the language in which the survey was filled as the current behavior is good I think (print the selected certifications in bulk in the current language).

I have removed the upgrade as you have proposed (detecting the survey languages and clearing it when it correspond to all installed language was hard to do in SQL for little gain). The survey are then migrated with lang_ids empty meaning that they can be started in all installed languages.

flch-odoo reviewed (commented):
Thank you for the update and answers, @pyduf!

flch-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -269,6 +273,23 @@ def get_print_url(self):
+    def _get_lang_with_fallback(self):
+        """ :return: the most suitable language for the user that is supported by the survey. """
+        self.ensure_one()
+        lang_code = self.lang_id.code or self.env.context.get('lang') or self.env['ir.http']._get_default_lang().code
```
As this only work with a `request`, shouldn't it be a controller method/util instead?

flch-odoo commented on addons/survey/models/__init__.py:
```diff
@@ -8,3 +8,4 @@
 from . import badge
 from . import challenge
 from . import res_partner
+from . import ir_http
```
order :hammer:  :slightly_smiling_face:  !

As we're in master, we could also reorder everything while we're here.

flch-odoo commented on addons/survey/models/ir_http.py:
```diff
@@ -0,0 +1,19 @@
+class IrHttp(models.AbstractModel):
+    _inherit = 'ir.http'
+
+    @api.model
+    def _is_survey_frontend(self, path):
+        return bool(re.match('^/survey/|^/[a-z]{2}/survey/|^/[a-z]{2}_[A-Z]{2}/survey/', path))
```
(I'd put it below the public method)

flch-odoo reviewed (commented):
Nice work, @pyduf! Could we get a review from @odoo/rd-security for using document.location and @odoo/rd-website for the last commit please?

pyduf reviewed (commented):
Hello @xmo-odoo and @flch-odoo , 

Thanks for the review, I have answered your comments and updated the code.

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,24 @@
+        """ When a language is disabled, clear corresponding survey language. """
+        if 'active' in vals and not vals['active']:
+            self.env['survey.user_input'].search([('lang_id', 'in', self.ids)]).lang_id = False
+            surveys = self.env['survey.survey'].search([('lang_ids', 'in', self.ids)])
```
```suggestion
            surveys_sudo = self.env['survey.survey'].sudo().search([('lang_ids', 'in', self.ids)])
```

pyduf replied:
Nice catch, thanks.

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,24 @@
+            surveys_all_langs = surveys.filtered(lambda survey: not survey.lang_ids)
+            surveys.write({'lang_ids': [Command.unlink(lang.id) for lang in self]})
+            survey_all_langs_after_disabling_lang = surveys.filtered(lambda survey: not survey.lang_ids)
+            if surveys_all_langs != survey_all_langs_after_disabling_lang:
+                impacted_surveys = survey_all_langs_after_disabling_lang - surveys_all_langs
```
```suggestion
            if impacted_surveys := survey_all_langs_after_disabling_lang - surveys_all_langs:
```

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,24 @@
+                raise UserError(
+                    _("Cannot deactivate a language that is currently used by a survey supporting only that language"
+                      " (surveys: %(survey_titles)s).",
+                      survey_titles=', '.join([f'"{survey.title})"' for survey in impacted_surveys])))
```
Show names only if read access_right  on all of them.

pyduf replied:
Yes indeed, thanks.

flch-odoo commented on addons/survey/models/survey_survey.py:
```diff
@@ -43,6 +43,13 @@ def default_get(self, fields_list):
         string='Survey Type', required=True, default='custom')
+    lang_ids = fields.Many2many(
+        'res.lang', string='Languages',
+        default=lambda self: self.env['res.lang']._lang_get(
+            self.env.context.get('lang') or self.env['res.lang'].get_installed()[0][0]),
```
Checking if we should do something to have `False` at upgrade. 

@filisbits :)

pyduf replied:
Nice catch also, thanks.

pyduf reviewed (commented):
Hello @flch-odoo , 

I have made the changes. Thanks for your advice and your help on this task.

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,26 @@
+        if 'active' in vals and not vals['active']:
+            self.env['survey.user_input'].search([('lang_id', 'in', self.ids)]).lang_id = False
+            surveys_sudo = self.env['survey.survey'].search([('lang_ids', 'in', self.ids)]).sudo()
+            surveys_sudo_all_langs = surveys_sudo.filtered(lambda survey: not survey.lang_ids)
```
Are those even possible since you searched those with the `lang`?

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,26 @@
+        """ When a language is disabled, clear corresponding survey language. """
+        if 'active' in vals and not vals['active']:
+            self.env['survey.user_input'].search([('lang_id', 'in', self.ids)]).lang_id = False
+            surveys_sudo = self.env['survey.survey'].search([('lang_ids', 'in', self.ids)]).sudo()
```
Or you don't find all records, do you?
```suggestion
            surveys_sudo = self.env['survey.survey'].sudo().search([('lang_ids', 'in', self.ids)])
```

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,26 @@
+            surveys_sudo_all_langs = surveys_sudo.filtered(lambda survey: not survey.lang_ids)
+            surveys_sudo.write({'lang_ids': [Command.unlink(lang.id) for lang in self]})
+            survey_sudo_all_langs_after_disabling_lang = surveys_sudo.filtered(lambda survey: not survey.lang_ids)
+            if impacted_surveys_sudo := survey_sudo_all_langs_after_disabling_lang - surveys_sudo_all_langs:
```
Isn't it simply filtering `surveys_sudo.filtered(lambda s: len(s.lang_ids) == len(self))`, so that we could even check before `write`?

flch-odoo commented on addons/survey/models/res_lang.py:
```diff
@@ -0,0 +1,26 @@
+                if self.env['survey.survey'].search([('id', 'in', impacted_surveys_sudo.ids)]) == impacted_surveys_sudo:
+                    error += '\n'
+                    error += _("Survey(s): %(surveys_list)s",
+                               surveys_list=', '.join([f'"{survey.title}"' for survey in impacted_surveys_sudo]))
```
```suggestion
                               surveys_list=', '.join(f'"{survey.title}"' for survey in impacted_surveys_sudo))
```

flch-odoo commented on addons/survey/controllers/main.py:
```diff
@@ -874,3 +888,23 @@ def _prepare_search_filter_input_line(self, user_input_line):
+        user_input.ensure_one()
+        user_input_sudo = user_input.sudo()
+        if user_input_sudo.lang_id:
+            return user_input_sudo.lang_id
```
```suggestion
            return user_input_sudo.lang_id.sudo(False)
```

flch-odoo reviewed (commented):
Thanks for the thorough work, @pyduf 
Just a few small polishing+sudo related remarks, we have the time to handle these.

flch-odoo commented:
@robodoo r+

robodoo commented:
@pyduf @flch-odoo because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

flch-odoo commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@pyduf @flch-odoo linked pull request(s) odoo/upgrade#7228 not ready. Linked PRs are not staged until all of them are ready.

pyduf reviewed (commented):
Hello @flch-odoo ,

Thanks for the useful comments, I have made the corrections.
